src/modalities/models/gpt2/gpt2_model.py

```diff
@@ -1,18 +1,116 @@
-import math
+import logging
 from enum import Enum
+
+import math
 from functools import partial
-from typing import Annotated, Dict
+from typing import Annotated, Dict, List, Tuple
 
 import torch
 import torch.nn as nn
 import xformers.ops as xops
-from pydantic import BaseModel, Field, model_validator
+from pydantic import BaseModel, Field, model_validator, validator
 from torch.nn import functional as F
+from xformers.components.positional_embedding import RotaryEmbedding
 
+from modalities.config.utils import convert_base_model_config_to_dict
 from modalities.models.model import NNModel
+from modalities.util import parse_enum_by_name
+
 
 # GPT2 implementation taken from nanogpt https://github.com/karpathy/nanoGPT
 
+class PositionTypes(str, Enum):
+    ABSOLUTE = "ABSOLUTE"
+    NOPE = "NOPE"
+
+
+class QueryKeyValueTransform(nn.Module):
```

Comment:
It would make sense, to turn this into an abstract class with and abstract forward() method.